services/src/main/scala/cromwell/pipeline/service/RunService.scala

```diff
@@ -38,36 +48,69 @@ object RunService {
+          .flatMap(_ => runRepository.addRun(newRun))
       }
 
       def getRunByIdAndUser(runId: RunId, projectId: ProjectId, userId: UserId): Future[Option[Run]] =
-        projectService.getUserProjectById(projectId, userId).flatMap { _ =>
-          runRepository.getRunByIdAndUser(runId, userId).map(_.filter(_.projectId == projectId))
-        }
+        projectService
+          .getUserProjectById(projectId, userId)
+          .recoverWith {
+            case e: ProjectServiceException => serviceErrorMapper(e)
+          }
+          .flatMap { _ =>
+            runRepository
+              .getRunByIdAndUser(runId, userId)
+              .recoverWith {
+                case _ => internalError("find run")
+              }
+              .map(_.filter(_.projectId == projectId))
+          }
 
       def getRunsByProject(projectId: ProjectId, userId: UserId): Future[Seq[Run]] =
-        projectService.getUserProjectById(projectId, userId).flatMap(_ => runRepository.getRunsByProject(projectId))
+        projectService
+          .getUserProjectById(projectId, userId)
+          .recoverWith {
+            case e: ProjectServiceException => serviceErrorMapper(e)
+          }
+          .flatMap(_ => runRepository.getRunsByProject(projectId).recoverWith { case _ => internalError("find run") })
 
       def deleteRunById(runId: RunId, projectId: ProjectId, userId: UserId): Future[Int] =
         getRunByIdAndUser(runId, projectId, userId).flatMap {
-          case Some(_) => runRepository.deleteRunById(runId)
-          case None    => Future.failed(new RuntimeException("run with this id doesn't exist"))
+          case Some(_) => runRepository.deleteRunById(runId).recoverWith { case _ => internalError("delete run") }
+          case None    => Future.failed(NotFound())
         }
 
       def updateRun(runId: RunId, request: RunUpdateRequest, projectId: ProjectId, userId: UserId): Future[Int] =
         getRunByIdAndUser(runId, projectId, userId).flatMap {
           case Some(run) =>
-            runRepository.updateRun(
-              run.copy(
-                status = request.status,
-                timeStart = request.timeStart,
-                timeEnd = request.timeEnd,
-                results = request.results,
-                cmwlWorkflowId = request.cmwlWorkflowId
+            runRepository
+              .updateRun(
+                run.copy(
+                  status = request.status,
+                  timeStart = request.timeStart,
+                  timeEnd = request.timeEnd,
+                  results = request.results,
+                  cmwlWorkflowId = request.cmwlWorkflowId
+                )
               )
-            )
-          case None => Future.failed(new RuntimeException("run with this id doesn't exist"))
+              .recoverWith { case _ => internalError("update run") }
+          case None => Future.failed(NotFound())
+        }
+
+      private def internalError(action: String): Future[Nothing] =
```

Comment:
That method name does not look like an action. Maybe "getInternalError" or smth like that would be better

Reply:
This name reflects the results you get and is familiar. We do not pass the result outside the class boundaries. In addition, in other classes, it is also named the same, I think it is not worth changing traditions :)